# 01 - Limpeza e Tratamentos dos dados


## Primeiros passos

In [ ]:
#Importa bibliotecas utilizadas
import pandas as pd
import numpy as np
import json

In [ ]:
#Abre o arquivo
with open('.../Dados/Telco-Customer-Churn.json') as file:
  data = json.load(file)

In [ ]:
#Mostra o arquivo normalizado
df = pd.json_normalize(data)
df

customerID Churn customer.gender  customer.SeniorCitizen  \
0     0002-ORFBO    No          Female                       0   
1     0003-MKNFE    No            Male                       0   
2     0004-TLHLJ   Yes            Male                       0   
3     0011-IGKFF   Yes            Male                       1   
4     0013-EXCHZ   Yes          Female                       1   
...          ...   ...             ...                     ...   
7262  9987-LUTYD    No          Female                       0   
7263  9992-RRAMN   Yes            Male                       0   
7264  9992-UJOEL    No            Male                       0   
7265  9993-LHIEB    No            Male                       0   
7266  9995-HOTOH    No            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
0                 Yes                 Yes                9                Yes   
1                  No                  No                9                Yes   
2                  No                  No                4                Yes   
3                 Yes                  No               13                Yes   
4                 Yes                  No                3                Yes   
...               ...                 ...              ...                ...   
7262               No                  No               13                Yes   
7263              Yes                  No               22                Yes   
7264               No                  No                2                Yes   
7265              Yes                 Yes               67                Yes   
7266              Yes                 Yes               63                 No   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
0                     No                      DSL  ...                   Yes   
1                    Yes                      DSL  ...                    No   
2                     No              Fiber optic  ...                    No   
3                     No              Fiber optic  ...                   Yes   
4                     No              Fiber optic  ...                    No   
...                  ...                      ...  ...                   ...   
7262                  No                      DSL  ...                    No   
7263                 Yes              Fiber optic  ...                    No   
7264                  No                      DSL  ...                   Yes   
7265                  No                      DSL  ...                    No   
7266    No phone service                      DSL  ...                   Yes   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
0                           No                  Yes                  Yes   
1                           No                   No                   No   
2                          Yes                   No                   No   
3                          Yes                   No                  Yes   
4                           No                  Yes                  Yes   
...                        ...                  ...                  ...   
7262                        No                  Yes                   No   
7263                        No                   No                   No   
7264                        No                   No                   No   
7265                       Yes                  Yes                   No   
7266                       Yes                   No                  Yes   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
0                          No         One year                      Yes   
1                         Yes   Month-to-month                       No   
2                          No   Month-to-month                      Yes   
3                         Yes   Month-to-month                    

## Verificando se existem valores nulos


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7267 non-null   object 
 1   Churn                      7267 non-null   object 
 2   customer.gender            7267 non-null   object 
 3   customer.SeniorCitizen     7267 non-null   int64  
 4   customer.Partner           7267 non-null   object 
 5   customer.Dependents        7267 non-null   object 
 6   customer.tenure            7267 non-null   int64  
 7   phone.PhoneService         7267 non-null   object 
 8   phone.MultipleLines        7267 non-null   object 
 9   internet.InternetService   7267 non-null   object 
 10  internet.OnlineSecurity    7267 non-null   object 
 11  internet.OnlineBackup      7267 non-null   object 
 12  internet.DeviceProtection  7267 non-null   object 
 13  internet.TechSupport       7267 non-null   objec

Pode-se observar que não há valores nulos em nenhuma coluna do Data Frame

## Outras Verificações:

O próximo passo é analisar cada uma das colunas individualmente com o objetivo de:
+ Verificar se existem valores em branco 
+ Verificar se respostas correspondem com respectivas colunas 



In [ ]:
df.isna().sum()

customerID                   0
Churn                        0
customer.gender              0
customer.SeniorCitizen       0
customer.Partner             0
customer.Dependents          0
customer.tenure              0
phone.PhoneService           0
phone.MultipleLines          0
internet.InternetService     0
internet.OnlineSecurity      0
internet.OnlineBackup        0
internet.DeviceProtection    0
internet.TechSupport         0
internet.StreamingTV         0
internet.StreamingMovies     0
account.Contract             0
account.PaperlessBilling     0
account.PaymentMethod        0
account.Charges.Monthly      0
account.Charges.Total        0
dtype: int64

Percebe-se que não há valores em branco, aparentemente.

In [ ]:
#Imprime valores contidos em cada coluna
for coluna in df.columns:
    print(df[coluna].value_counts())

0002-ORFBO    1
6614-VBEGU    1
6637-KYRCV    1
6635-MYYYZ    1
6635-CPNUN    1
             ..
3374-TTZTK    1
3374-PZLXD    1
3374-LXDEV    1
3373-YZZYM    1
9995-HOTOH    1
Name: customerID, Length: 7267, dtype: int64
No     5174
Yes    1869
        224
Name: Churn, dtype: int64
Male      3675
Female    3592
Name: customer.gender, dtype: int64
0    6085
1    1182
Name: customer.SeniorCitizen, dtype: int64
No     3749
Yes    3518
Name: customer.Partner, dtype: int64
No     5086
Yes    2181
Name: customer.Dependents, dtype: int64
1     634
72    369
2     246
3     207
4     185
     ... 
38     60
39     59
44     54
36     50
0      11
Name: customer.tenure, Length: 73, dtype: int64
Yes    6560
No      707
Name: phone.PhoneService, dtype: int64
No                  3495
Yes                 3065
No phone service     707
Name: phone.MultipleLines, dtype: int64
Fiber optic    3198
DSL            2488
No             1581
Name: internet.InternetService, dtype: int64
No                    

Com o método .value_counts() percebe-se que as os valores estão adequados ao que representa cada coluna do DataFrame, porém a coluna "Chur"apresenta valores em branco.

In [ ]:
df.Churn.unique() 

array(['No', 'Yes', ''], dtype=object)

In [ ]:
#Seleciona as linhas com Churn em branco
df.query("Churn == ''") 

customerID Churn customer.gender  customer.SeniorCitizen  \
30    0047-ZHDTW                Female                       0   
75    0120-YZLQA                  Male                       0   
96    0154-QYHJU                  Male                       0   
98    0162-RZGMZ                Female                       1   
175   0274-VVQOQ                  Male                       1   
...          ...   ...             ...                     ...   
7158  9840-GSRFX                Female                       0   
7180  9872-RZQQB                Female                       0   
7211  9920-GNDMB                  Male                       0   
7239  9955-RVWSC                Female                       0   
7247  9966-VYRTZ                Female                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
30                 No                  No               11                Yes   
75                 No                  No               71                Yes   
96                 No                  No               29                Yes   
98                 No                  No                5                Yes   
175               Yes                  No               65                Yes   
...               ...                 ...              ...                ...   
7158               No                  No               14                Yes   
7180              Yes                  No               49                 No   
7211               No                  No                9                Yes   
7239              Yes                 Yes               67                Yes   
7247              Yes                 Yes               31                Yes   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
30                   Yes              Fiber optic  ...                    No   
75                    No                       No  ...   No internet service   
96                    No                      DSL  ...                   Yes   
98                    No                      DSL  ...                   Yes   
175                  Yes              Fiber optic  ...                   Yes   
...                  ...                      ...  ...                   ...   
7158                 Yes                      DSL  ...                   Yes   
7180    No phone service                      DSL  ...                    No   
7211                 Yes              Fiber optic  ...                    No   
7239                  No                       No  ...   No internet service   
7247                  No                       No  ...   No internet service   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
30                          No                   No                   No   
75         No internet service  No internet service  No internet service   
96                          No                  Yes                   No   
98                          No                  Yes                   No   
175                        Yes                   No                  Yes   
...                        ...                  ...                  ...   
7158                        No                   No                   No   
7180                        No                   No                  Yes   
7211                        No                   No                   No   
7239       No internet service  No internet service  No internet service   
7247       No internet service  No internet service  No internet service   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
30                         No   Month-to-month                      Yes   
75        No internet service         Two year                      Yes   
96                         No         One year                      Yes   
98                         No   Month-to-month                    

Temos a confirmação de que 224 linhas possuem valores em branco na coluna Churn. Uma vez que esta é a variável mais importante, ou seja, é a coluna que contém os resultados que queremos classificar ao final do problema, foi optado por excluir estas 224 linhas.

In [ ]:
#Cria filtro que exclui as linhas com Churn em branco 
df = df.query("Churn!= ''") 
df

customerID Churn customer.gender  customer.SeniorCitizen  \
0     0002-ORFBO    No          Female                       0   
1     0003-MKNFE    No            Male                       0   
2     0004-TLHLJ   Yes            Male                       0   
3     0011-IGKFF   Yes            Male                       1   
4     0013-EXCHZ   Yes          Female                       1   
...          ...   ...             ...                     ...   
7262  9987-LUTYD    No          Female                       0   
7263  9992-RRAMN   Yes            Male                       0   
7264  9992-UJOEL    No            Male                       0   
7265  9993-LHIEB    No            Male                       0   
7266  9995-HOTOH    No            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
0                 Yes                 Yes                9                Yes   
1                  No                  No                9                Yes   
2                  No                  No                4                Yes   
3                 Yes                  No               13                Yes   
4                 Yes                  No                3                Yes   
...               ...                 ...              ...                ...   
7262               No                  No               13                Yes   
7263              Yes                  No               22                Yes   
7264               No                  No                2                Yes   
7265              Yes                 Yes               67                Yes   
7266              Yes                 Yes               63                 No   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
0                     No                      DSL  ...                   Yes   
1                    Yes                      DSL  ...                    No   
2                     No              Fiber optic  ...                    No   
3                     No              Fiber optic  ...                   Yes   
4                     No              Fiber optic  ...                    No   
...                  ...                      ...  ...                   ...   
7262                  No                      DSL  ...                    No   
7263                 Yes              Fiber optic  ...                    No   
7264                  No                      DSL  ...                   Yes   
7265                  No                      DSL  ...                    No   
7266    No phone service                      DSL  ...                   Yes   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
0                           No                  Yes                  Yes   
1                           No                   No                   No   
2                          Yes                   No                   No   
3                          Yes                   No                  Yes   
4                           No                  Yes                  Yes   
...                        ...                  ...                  ...   
7262                        No                  Yes                   No   
7263                        No                   No                   No   
7264                        No                   No                   No   
7265                       Yes                  Yes                   No   
7266                       Yes                   No                  Yes   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
0                          No         One year                      Yes   
1                         Yes   Month-to-month                       No   
2                          No   Month-to-month                      Yes   
3                         Yes   Month-to-month                    

In [ ]:
df.Churn.unique()

array(['No', 'Yes'], dtype=object)

A coluna "Charges Total" está sendo lida como typo objetc, ao invés de float. Portanto vamos transformar estes valores em numéricos

In [ ]:
df['account.Charges.Total'].unique()

array(['593.3', '542.4', '280.85', ..., '742.9', '4627.65', '3707.6'],
      dtype=object)

In [ ]:
#Transforma account.Charges.Total em valores numericos
#df["account.Charges.Total"] = pd.to_numeric(df["account.Charges.Total"])          #erro na linha 945

Aqui os valores estão sendo percebidos como "object", porém ao tentar transformar-los em valores float, encontramos um erro pois, algumas linhas possuem valores em branco na coluna "Charges Total".

In [ ]:
#Analisa os dados contidos na linha em que ocorreu erro
df.iloc[945]

customerID                                1371-DWPAZ
Churn                                             No
customer.gender                               Female
customer.SeniorCitizen                             0
customer.Partner                                 Yes
customer.Dependents                              Yes
customer.tenure                                    0
phone.PhoneService                                No
phone.MultipleLines                 No phone service
internet.InternetService                         DSL
internet.OnlineSecurity                          Yes
internet.OnlineBackup                            Yes
internet.DeviceProtection                        Yes
internet.TechSupport                             Yes
internet.StreamingTV                             Yes
internet.StreamingMovies                          No
account.Contract                            Two year
account.PaperlessBilling                          No
account.PaymentMethod        Credit card (auto

In [ ]:
df.isna().sum()

customerID                   0
Churn                        0
customer.gender              0
customer.SeniorCitizen       0
customer.Partner             0
customer.Dependents          0
customer.tenure              0
phone.PhoneService           0
phone.MultipleLines          0
internet.InternetService     0
internet.OnlineSecurity      0
internet.OnlineBackup        0
internet.DeviceProtection    0
internet.TechSupport         0
internet.StreamingTV         0
internet.StreamingMovies     0
account.Contract             0
account.PaperlessBilling     0
account.PaymentMethod        0
account.Charges.Monthly      0
account.Charges.Total        0
dtype: int64

In [ ]:
#Procura por outras linhas que possuem account.Charges.Total em branco
df.loc[(df['account.Charges.Total'] == ' ')] 

customerID Churn customer.gender  customer.SeniorCitizen  \
975   1371-DWPAZ    No          Female                       0   
1775  2520-SGTTA    No          Female                       0   
1955  2775-SEFEE    No            Male                       0   
2075  2923-ARZLG    No            Male                       0   
2232  3115-CZMZD    No            Male                       0   
2308  3213-VVOLG    No            Male                       0   
2930  4075-WKNIU    No          Female                       0   
3134  4367-NUYAO    No            Male                       0   
3203  4472-LVYGI    No          Female                       0   
4169  5709-LVOEQ    No          Female                       0   
5599  7644-OMVMY    No            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
975               Yes                 Yes                0                 No   
1775              Yes                 Yes                0                Yes   
1955               No                 Yes                0                Yes   
2075              Yes                 Yes                0                Yes   
2232               No                 Yes                0                Yes   
2308              Yes                 Yes                0                Yes   
2930              Yes                 Yes                0                Yes   
3134              Yes                 Yes                0                Yes   
3203              Yes                 Yes                0                 No   
4169              Yes                 Yes                0                Yes   
5599              Yes                 Yes                0                Yes   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
975     No phone service                      DSL  ...                   Yes   
1775                  No                       No  ...   No internet service   
1955                 Yes                      DSL  ...                   Yes   
2075                  No                       No  ...   No internet service   
2232                  No                       No  ...   No internet service   
2308                 Yes                       No  ...   No internet service   
2930                 Yes                      DSL  ...                   Yes   
3134                 Yes                       No  ...   No internet service   
3203    No phone service                      DSL  ...                    No   
4169                  No                      DSL  ...                   Yes   
5599                  No                       No  ...   No internet service   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
975                        Yes                  Yes                  Yes   
1775       No internet service  No internet service  No internet service   
1955                        No                  Yes                   No   
2075       No internet service  No internet service  No internet service   
2232       No internet service  No internet service  No internet service   
2308       No internet service  No internet service  No internet service   
2930                       Yes                  Yes                  Yes   
3134       No internet service  No internet service  No internet service   
3203                       Yes                  Yes                  Yes   
4169                       Yes                   No                  Yes   
5599       No internet service  No internet service  No internet service   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
975                        No         Two year                       No   
1775      No internet service         Two year                       No   
1955                       No         Two year                      Yes   
2075      No internet service         One year                    

Após alguns testes, foi possível perceber que  todas as 11 linhas que possuem "Charges.Total" em branco também possuem a coluna "customer.ternure" = 0 

In [ ]:
#Procura por linhas que possuem customer.tenure em branco
df.loc[(df['customer.tenure'] == 0)] 

customerID Churn customer.gender  customer.SeniorCitizen  \
975   1371-DWPAZ    No          Female                       0   
1775  2520-SGTTA    No          Female                       0   
1955  2775-SEFEE    No            Male                       0   
2075  2923-ARZLG    No            Male                       0   
2232  3115-CZMZD    No            Male                       0   
2308  3213-VVOLG    No            Male                       0   
2930  4075-WKNIU    No          Female                       0   
3134  4367-NUYAO    No            Male                       0   
3203  4472-LVYGI    No          Female                       0   
4169  5709-LVOEQ    No          Female                       0   
5599  7644-OMVMY    No            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
975               Yes                 Yes                0                 No   
1775              Yes                 Yes                0                Yes   
1955               No                 Yes                0                Yes   
2075              Yes                 Yes                0                Yes   
2232               No                 Yes                0                Yes   
2308              Yes                 Yes                0                Yes   
2930              Yes                 Yes                0                Yes   
3134              Yes                 Yes                0                Yes   
3203              Yes                 Yes                0                 No   
4169              Yes                 Yes                0                Yes   
5599              Yes                 Yes                0                Yes   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
975     No phone service                      DSL  ...                   Yes   
1775                  No                       No  ...   No internet service   
1955                 Yes                      DSL  ...                   Yes   
2075                  No                       No  ...   No internet service   
2232                  No                       No  ...   No internet service   
2308                 Yes                       No  ...   No internet service   
2930                 Yes                      DSL  ...                   Yes   
3134                 Yes                       No  ...   No internet service   
3203    No phone service                      DSL  ...                    No   
4169                  No                      DSL  ...                   Yes   
5599                  No                       No  ...   No internet service   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
975                        Yes                  Yes                  Yes   
1775       No internet service  No internet service  No internet service   
1955                        No                  Yes                   No   
2075       No internet service  No internet service  No internet service   
2232       No internet service  No internet service  No internet service   
2308       No internet service  No internet service  No internet service   
2930                       Yes                  Yes                  Yes   
3134       No internet service  No internet service  No internet service   
3203                       Yes                  Yes                  Yes   
4169                       Yes                   No                  Yes   
5599       No internet service  No internet service  No internet service   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
975                        No         Two year                       No   
1775      No internet service         Two year                       No   
1955                       No         Two year                      Yes   
2075      No internet service         One year                    

E o inverso também se mostra verdadeiro, ou seja,  todas as linhas que possuem o customer.ternure = 0 possuem Charges.Total em branco.

Entende-se, portanto, que os valores estão em branco pois os contratos duraram menos de 1 mês. Sendo assim, a proxima linha de código iguala os valores em brancos em "account.Charges.Total" aos valores da coluna "account.Charges.Monthly", que neste caso, representa o valor real cobrado.

In [ ]:
df.loc[(df['account.Charges.Total'] == ' '),'account.Charges.Total'] = df.loc[(df['account.Charges.Total'] == ' '),'account.Charges.Monthly']

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:1773: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)


In [ ]:
df

customerID Churn customer.gender  customer.SeniorCitizen  \
0     0002-ORFBO    No          Female                       0   
1     0003-MKNFE    No            Male                       0   
2     0004-TLHLJ   Yes            Male                       0   
3     0011-IGKFF   Yes            Male                       1   
4     0013-EXCHZ   Yes          Female                       1   
...          ...   ...             ...                     ...   
7262  9987-LUTYD    No          Female                       0   
7263  9992-RRAMN   Yes            Male                       0   
7264  9992-UJOEL    No            Male                       0   
7265  9993-LHIEB    No            Male                       0   
7266  9995-HOTOH    No            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
0                 Yes                 Yes                9                Yes   
1                  No                  No                9                Yes   
2                  No                  No                4                Yes   
3                 Yes                  No               13                Yes   
4                 Yes                  No                3                Yes   
...               ...                 ...              ...                ...   
7262               No                  No               13                Yes   
7263              Yes                  No               22                Yes   
7264               No                  No                2                Yes   
7265              Yes                 Yes               67                Yes   
7266              Yes                 Yes               63                 No   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
0                     No                      DSL  ...                   Yes   
1                    Yes                      DSL  ...                    No   
2                     No              Fiber optic  ...                    No   
3                     No              Fiber optic  ...                   Yes   
4                     No              Fiber optic  ...                    No   
...                  ...                      ...  ...                   ...   
7262                  No                      DSL  ...                    No   
7263                 Yes              Fiber optic  ...                    No   
7264                  No                      DSL  ...                   Yes   
7265                  No                      DSL  ...                    No   
7266    No phone service                      DSL  ...                   Yes   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
0                           No                  Yes                  Yes   
1                           No                   No                   No   
2                          Yes                   No                   No   
3                          Yes                   No                  Yes   
4                           No                  Yes                  Yes   
...                        ...                  ...                  ...   
7262                        No                  Yes                   No   
7263                        No                   No                   No   
7264                        No                   No                   No   
7265                       Yes                  Yes                   No   
7266                       Yes                   No                  Yes   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
0                          No         One year                      Yes   
1                         Yes   Month-to-month                       No   
2                          No   Month-to-month                      Yes   
3                         Yes   Month-to-month                    

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7043 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7043 non-null   object 
 1   Churn                      7043 non-null   object 
 2   customer.gender            7043 non-null   object 
 3   customer.SeniorCitizen     7043 non-null   int64  
 4   customer.Partner           7043 non-null   object 
 5   customer.Dependents        7043 non-null   object 
 6   customer.tenure            7043 non-null   int64  
 7   phone.PhoneService         7043 non-null   object 
 8   phone.MultipleLines        7043 non-null   object 
 9   internet.InternetService   7043 non-null   object 
 10  internet.OnlineSecurity    7043 non-null   object 
 11  internet.OnlineBackup      7043 non-null   object 
 12  internet.DeviceProtection  7043 non-null   object 
 13  internet.TechSupport       7043 non-null   objec

Por fim, é possível transformar a coluna 20 para o tipo *float*

In [ ]:
df["account.Charges.Total"] = pd.to_numeric(df["account.Charges.Total"])  

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7043 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7043 non-null   object 
 1   Churn                      7043 non-null   object 
 2   customer.gender            7043 non-null   object 
 3   customer.SeniorCitizen     7043 non-null   int64  
 4   customer.Partner           7043 non-null   object 
 5   customer.Dependents        7043 non-null   object 
 6   customer.tenure            7043 non-null   int64  
 7   phone.PhoneService         7043 non-null   object 
 8   phone.MultipleLines        7043 non-null   object 
 9   internet.InternetService   7043 non-null   object 
 10  internet.OnlineSecurity    7043 non-null   object 
 11  internet.OnlineBackup      7043 non-null   object 
 12  internet.DeviceProtection  7043 non-null   object 
 13  internet.TechSupport       7043 non-null   objec

In [ ]:
df

customerID Churn customer.gender  customer.SeniorCitizen  \
0     0002-ORFBO    No          Female                       0   
1     0003-MKNFE    No            Male                       0   
2     0004-TLHLJ   Yes            Male                       0   
3     0011-IGKFF   Yes            Male                       1   
4     0013-EXCHZ   Yes          Female                       1   
...          ...   ...             ...                     ...   
7262  9987-LUTYD    No          Female                       0   
7263  9992-RRAMN   Yes            Male                       0   
7264  9992-UJOEL    No            Male                       0   
7265  9993-LHIEB    No            Male                       0   
7266  9995-HOTOH    No            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
0                 Yes                 Yes                9                Yes   
1                  No                  No                9                Yes   
2                  No                  No                4                Yes   
3                 Yes                  No               13                Yes   
4                 Yes                  No                3                Yes   
...               ...                 ...              ...                ...   
7262               No                  No               13                Yes   
7263              Yes                  No               22                Yes   
7264               No                  No                2                Yes   
7265              Yes                 Yes               67                Yes   
7266              Yes                 Yes               63                 No   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
0                     No                      DSL  ...                   Yes   
1                    Yes                      DSL  ...                    No   
2                     No              Fiber optic  ...                    No   
3                     No              Fiber optic  ...                   Yes   
4                     No              Fiber optic  ...                    No   
...                  ...                      ...  ...                   ...   
7262                  No                      DSL  ...                    No   
7263                 Yes              Fiber optic  ...                    No   
7264                  No                      DSL  ...                   Yes   
7265                  No                      DSL  ...                    No   
7266    No phone service                      DSL  ...                   Yes   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
0                           No                  Yes                  Yes   
1                           No                   No                   No   
2                          Yes                   No                   No   
3                          Yes                   No                  Yes   
4                           No                  Yes                  Yes   
...                        ...                  ...                  ...   
7262                        No                  Yes                   No   
7263                        No                   No                   No   
7264                        No                   No                   No   
7265                       Yes                  Yes                   No   
7266                       Yes                   No                  Yes   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
0                          No         One year                      Yes   
1                         Yes   Month-to-month                       No   
2                          No   Month-to-month                      Yes   
3                         Yes   Month-to-month                    

## Aplicando Econding nas variáveis
Agora, iremos transformas os valores *object* em valores numéricos binários. Desta forma, o processo de análise dos dados, bem como a criação do modelo de Machine Learnng serão facilitados. 

Para as variáveis que possuem respostas somente "Yes" e "No", vamos apenas substituir por 1 e 0.

In [ ]:
for coluna in df.columns:
    print(df[coluna].value_counts())

0002-ORFBO    1
6616-AALSR    1
6625-UTXEW    1
6625-IUTTT    1
6625-FLENO    1
             ..
3352-RICWQ    1
3352-ALMCK    1
3351-NQLDI    1
3351-NGXYI    1
9995-HOTOH    1
Name: customerID, Length: 7043, dtype: int64
No     5174
Yes    1869
Name: Churn, dtype: int64
Male      3555
Female    3488
Name: customer.gender, dtype: int64
0    5901
1    1142
Name: customer.SeniorCitizen, dtype: int64
No     3641
Yes    3402
Name: customer.Partner, dtype: int64
No     4933
Yes    2110
Name: customer.Dependents, dtype: int64
1     613
72    362
2     238
3     200
4     176
     ... 
28     57
39     56
44     51
36     50
0      11
Name: customer.tenure, Length: 73, dtype: int64
Yes    6361
No      682
Name: phone.PhoneService, dtype: int64
No                  3390
Yes                 2971
No phone service     682
Name: phone.MultipleLines, dtype: int64
Fiber optic    3096
DSL            2421
No             1526
Name: internet.InternetService, dtype: int64
No                     3498
Yes   

In [ ]:
colunas1 = ['Churn', 'customer.Partner', 'customer.Dependents', 'phone.PhoneService', 'account.PaperlessBilling']
for coluna in colunas1:
  df[coluna] = df[coluna].map({'No' : 0, 'Yes' : 1})

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [ ]:
df

customerID  Churn customer.gender  customer.SeniorCitizen  \
0     0002-ORFBO      0          Female                       0   
1     0003-MKNFE      0            Male                       0   
2     0004-TLHLJ      1            Male                       0   
3     0011-IGKFF      1            Male                       1   
4     0013-EXCHZ      1          Female                       1   
...          ...    ...             ...                     ...   
7262  9987-LUTYD      0          Female                       0   
7263  9992-RRAMN      1            Male                       0   
7264  9992-UJOEL      0            Male                       0   
7265  9993-LHIEB      0            Male                       0   
7266  9995-HOTOH      0            Male                       0   

      customer.Partner  customer.Dependents  customer.tenure  \
0                    1                    1                9   
1                    0                    0                9   
2                    0                    0                4   
3                    1                    0               13   
4                    1                    0                3   
...                ...                  ...              ...   
7262                 0                    0               13   
7263                 1                    0               22   
7264                 0                    0                2   
7265                 1                    1               67   
7266                 1                    1               63   

      phone.PhoneService phone.MultipleLines internet.InternetService  ...  \
0                      1                  No                      DSL  ...   
1                      1                 Yes                      DSL  ...   
2                      1                  No              Fiber optic  ...   
3                      1                  No              Fiber optic  ...   
4                      1                  No              Fiber optic  ...   
...                  ...                 ...                      ...  ...   
7262                   1                  No                      DSL  ...   
7263                   1                 Yes              Fiber optic  ...   
7264                   1                  No                      DSL  ...   
7265                   1                  No                      DSL  ...   
7266                   0    No phone service                      DSL  ...   

     internet.OnlineBackup internet.DeviceProtection internet.TechSupport  \
0                      Yes                        No                  Yes   
1                       No                        No                   No   
2                       No                       Yes                   No   
3                      Yes                       Yes                   No   
4                       No                        No                  Yes   
...                    ...                       ...                  ...   
7262                    No                        No                  Yes   
7263                    No                        No                   No   
7264                   Yes                        No                   No   
7265                    No                       Yes                  Yes   
7266                   Yes                       Yes                   No   

     internet.StreamingTV internet.StreamingMovies account.Contract  \
0                     Yes                       No         One year   
1                      No                      Yes   Month-to-month   
2                      No                       No   Month-to-month   
3                     Yes                      Yes   Month-to-month   
4                     Yes                       No   Month-to-month   
...                   ...                      ...              ...   
7262                   No                       No         One year   
72

A coluna "account.Contract" vamos considerar como valores ordinais, variando entre 0, 1 ou 2 anos:

In [ ]:
df['account.Contract'] = df['account.Contract'].map({'Month-to-month' : 0, 'One year' : 1, 'Two year' : 2})
df.head()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


customerID  Churn customer.gender  customer.SeniorCitizen  \
0  0002-ORFBO      0          Female                       0   
1  0003-MKNFE      0            Male                       0   
2  0004-TLHLJ      1            Male                       0   
3  0011-IGKFF      1            Male                       1   
4  0013-EXCHZ      1          Female                       1   

   customer.Partner  customer.Dependents  customer.tenure  phone.PhoneService  \
0                 1                    1                9                   1   
1                 0                    0                9                   1   
2                 0                    0                4                   1   
3                 1                    0               13                   1   
4                 1                    0                3                   1   

  phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
0                  No                      DSL  ...                   Yes   
1                 Yes                      DSL  ...                    No   
2                  No              Fiber optic  ...                    No   
3                  No              Fiber optic  ...                   Yes   
4                  No              Fiber optic  ...                    No   

  internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
0                        No                  Yes                  Yes   
1                        No                   No                   No   
2                       Yes                   No                   No   
3                       Yes                   No                  Yes   
4                        No                  Yes                  Yes   

  internet.StreamingMovies account.Contract  account.PaperlessBilling  \
0                       No                1                         1   
1                      Yes                0                         0   
2                       No                0                         1   
3                      Yes                0                         1   
4                       No                0                         1   

   account.PaymentMethod account.Charges.Monthly  account.Charges.Total  
0           Mailed check                    65.6                 593.30  
1           Mailed check                    59.9                 542.40  
2       Electronic check                    73.9                 280.85  
3       Electronic check                    98.0                1237.85  
4           Mailed check                    83.9                 267.40  

[5 rows x 21 columns]

E no restante das colunas, vamos representar as respostas em formato binário, utilizando a função **get_dummies** do Pandas. Optou-se por usar esta função do Pandas ao invés de utilizar a função **OneHotEncoder** da biblioteca Sklearn, pois gostaríamos de manter o formato DataFrame do dados e trabalhar com os nomes das colunas, de forma que seja possível analisar as variáveis de forma visual e possíveis relações entre elas . Desta forma, poderemos gerar tirar *insights* antes de criar o modelo de ML. 

In [ ]:
colunas2 = ['customer.gender', 
            'phone.MultipleLines', 
            'internet.InternetService', 
            'internet.OnlineSecurity', 
            'internet.OnlineBackup', 
            'internet.DeviceProtection', 
            'internet.TechSupport', 
            'internet.StreamingTV', 
            'internet.StreamingMovies',  
            'account.PaymentMethod'
            ]
for coluna in colunas2:
  transforma = pd.get_dummies(df[coluna], prefix = coluna)
  df= pd.concat([df, transforma], axis=1).drop([coluna], axis=1)

In [ ]:
df.head()

customerID  Churn  customer.SeniorCitizen  customer.Partner  \
0  0002-ORFBO      0                       0                 1   
1  0003-MKNFE      0                       0                 0   
2  0004-TLHLJ      1                       0                 0   
3  0011-IGKFF      1                       1                 1   
4  0013-EXCHZ      1                       1                 1   

   customer.Dependents  customer.tenure  phone.PhoneService  account.Contract  \
0                    1                9                   1                 1   
1                    0                9                   1                 0   
2                    0                4                   1                 0   
3                    0               13                   1                 0   
4                    0                3                   1                 0   

   account.PaperlessBilling  account.Charges.Monthly  ...  \
0                         1                     65.6  ...   
1                         0                     59.9  ...   
2                         1                     73.9  ...   
3                         1                     98.0  ...   
4                         1                     83.9  ...   

   internet.StreamingTV_No  internet.StreamingTV_No internet service  \
0                        0                                         0   
1                        1                                         0   
2                        1                                         0   
3                        0                                         0   
4                        0                                         0   

   internet.StreamingTV_Yes  internet.StreamingMovies_No  \
0                         1                            1   
1                         0                            0   
2                         0                            1   
3                         1                            0   
4                         1                            1   

   internet.StreamingMovies_No internet service  internet.StreamingMovies_Yes  \
0                                             0                             0   
1                                             0                             1   
2                                             0                             0   
3                                             0                             1   
4                                             0                             0   

   account.PaymentMethod_Bank transfer (automatic)  \
0                                                0   
1                                                0   
2                                                0   
3                                                0   
4                                                0   

   account.PaymentMethod_Credit card (automatic)  \
0                                              0   
1                                              0   
2                                              0   
3                                              0   
4                                              0   

   account.PaymentMethod_Electronic check  account.PaymentMethod_Mailed check  
0                                       0                                   1  
1                                       0                                   1  
2                                       1                                   0  
3                                       1                                   0  
4                                       0                                   1  

[5 rows x 41 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7043 entries, 0 to 7266
Data columns (total 41 columns):
 #   Column                                           Non-Null Count  Dtype  
---  ------                                           --------------  -----  
 0   customerID                                       7043 non-null   object 
 1   Churn                                            7043 non-null   int64  
 2   customer.SeniorCitizen                           7043 non-null   int64  
 3   customer.Partner                                 7043 non-null   int64  
 4   customer.Dependents                              7043 non-null   int64  
 5   customer.tenure                                  7043 non-null   int64  
 6   phone.PhoneService                               7043 non-null   int64  
 7   account.Contract                                 7043 non-null   int64  
 8   account.PaperlessBilling                         7043 non-null   int64  
 9   account.Charges.Monthly       

Todas as colunas que representam informações referentes a "No phone service" e "No internet service" são redundantes, pois essa informação já está sendo representada nas colunas "phone.PhoneService" e "internet.InternetService_No", respectivamente. Abaixo temos um exemplo de como as duas colunas são perfeitamente opostas. 

In [ ]:
redundancia = [df['phone.PhoneService'], df['phone.MultipleLines_No phone service']]
pd.DataFrame(redundancia)

0     1     2     3     4     5     \
phone.PhoneService                       1     1     1     1     1     1   
phone.MultipleLines_No phone service     0     0     0     0     0     0   

                                      6     7     8     9     ...  7257  7258  \
phone.PhoneService                       1     1     1     1  ...     1     1   
phone.MultipleLines_No phone service     0     0     0     0  ...     0     0   

                                      7259  7260  7261  7262  7263  7264  \
phone.PhoneService                       1     1     1     1     1     1   
phone.MultipleLines_No phone service     0     0     0     0     0     0   

                                      7265  7266  
phone.PhoneService                       1     0  
phone.MultipleLines_No phone service     0     1  

[2 rows x 7043 columns]

O mesmo ocorre para as colunas com "Yes" e "No" de um mesmo serviço, são perfeitamente opostas e portanto, com informações redundantes. Segue o exemplo:

In [ ]:
redundancia2 = [df['internet.OnlineSecurity_Yes'], df['internet.OnlineSecurity_No']]
pd.DataFrame(redundancia)       

0     1     2     3     4     5     \
phone.PhoneService                       1     1     1     1     1     1   
phone.MultipleLines_No phone service     0     0     0     0     0     0   

                                      6     7     8     9     ...  7257  7258  \
phone.PhoneService                       1     1     1     1  ...     1     1   
phone.MultipleLines_No phone service     0     0     0     0  ...     0     0   

                                      7259  7260  7261  7262  7263  7264  \
phone.PhoneService                       1     1     1     1     1     1   
phone.MultipleLines_No phone service     0     0     0     0     0     0   

                                      7265  7266  
phone.PhoneService                       1     0  
phone.MultipleLines_No phone service     0     1  

[2 rows x 7043 columns]

Portanto, para deixar nosso dados mais leves, todas estas colunas com informações redundantes serão exluídas. 

In [ ]:
df = df.drop(['phone.MultipleLines_No phone service',
              'phone.MultipleLines_No',
              'internet.OnlineSecurity_No internet service',
              'internet.OnlineBackup_No internet service',
              'internet.DeviceProtection_No internet service',
              'internet.TechSupport_No internet service',
              'internet.StreamingTV_No internet service',
              'internet.StreamingMovies_No internet service',
              'internet.OnlineSecurity_No',
              'internet.OnlineBackup_No',
              'internet.DeviceProtection_No',
              'internet.TechSupport_No',
              'internet.StreamingTV_No',
              'internet.StreamingMovies_No'
              ], axis=1)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7043 entries, 0 to 7266
Data columns (total 27 columns):
 #   Column                                           Non-Null Count  Dtype  
---  ------                                           --------------  -----  
 0   customerID                                       7043 non-null   object 
 1   Churn                                            7043 non-null   int64  
 2   customer.SeniorCitizen                           7043 non-null   int64  
 3   customer.Partner                                 7043 non-null   int64  
 4   customer.Dependents                              7043 non-null   int64  
 5   customer.tenure                                  7043 non-null   int64  
 6   phone.PhoneService                               7043 non-null   int64  
 7   account.Contract                                 7043 non-null   int64  
 8   account.PaperlessBilling                         7043 non-null   int64  
 9   account.Charges.Monthly       

Desta forma, foi possível reduzir nossos dados de 40 para 26 colunas, sem perda de informação.

## Traduzindo o nome das colunas
Considerando que podem haver pessoas na equipe que não dominam o idioma Inglês, as colunas serão renomeadas e traduzidas para incluir todos os participantes do time.

In [ ]:
renomear = {'customerID' : 'ID',
            'Churn': 'Evasao',
            'customer.SeniorCitizen': 'Idoso',
            'customer.Partner' : 'Parceiro',
            'customer.Dependents' : 'Dependentes',
            'customer.tenure' : 'Meses_Contrato',
            'phone.PhoneService' : 'Servico_Telefone',
            'account.Contract' : 'Tipo_Contrato',
            'account.PaperlessBilling' : 'Fatura_Digital',
            'account.Charges.Monthly' : 'Valor_Mensal',
            'account.Charges.Total' : 'Valor_Total',
            'customer.gender_Female': 'Gen_Feminino',
            'customer.gender_Male' : 'Gen_Masculino',
            'phone.MultipleLines_Yes' : 'Multiplas_Linhas',
            'internet.InternetService_DSL' : 'Internet_DSL',
            'internet.InternetService_Fiber optic': 'Internet_FibraOtica',
            'internet.InternetService_No' : 'Sem_Servico_Internet',
            'internet.OnlineSecurity_Yes' : 'Seguranca_Online',
            'internet.OnlineBackup_Yes' : 'Backup_Online',
            'internet.DeviceProtection_Yes': 'Protecao_Dispositivo',
            'internet.TechSupport_Yes' : 'Suporte_Tecnico_Internet',
            'internet.StreamingTV_Yes' : 'TV_Cabo',
            'internet.StreamingMovies_Yes' : 'Streaming_Filmes',
            'account.PaymentMethod_Bank transfer (automatic)' : 'Transferencia_Bancaria',
            'account.PaymentMethod_Credit card (automatic)' : 'Cartao_Credito',
            'account.PaymentMethod_Electronic check' : 'Cheque_Eletronico',
            'account.PaymentMethod_Mailed check' : 'Cheque'
            }
df.rename(columns = renomear, inplace = True)

In [ ]:
df

ID  Cliente_Saiu  Idoso  Parceiro  Dependentes  Meses_Contrato  \
0     0002-ORFBO             0      0         1            1               9   
1     0003-MKNFE             0      0         0            0               9   
2     0004-TLHLJ             1      0         0            0               4   
3     0011-IGKFF             1      1         1            0              13   
4     0013-EXCHZ             1      1         1            0               3   
...          ...           ...    ...       ...          ...             ...   
7262  9987-LUTYD             0      0         0            0              13   
7263  9992-RRAMN             1      0         1            0              22   
7264  9992-UJOEL             0      0         0            0               2   
7265  9993-LHIEB             0      0         1            1              67   
7266  9995-HOTOH             0      0         1            1              63   

      Servico_Telefone  Tipo_Contrato  Fatura_Digital  Valor_Mensal  ...  \
0                    1              1               1         65.60  ...   
1                    1              0               0         59.90  ...   
2                    1              0               1         73.90  ...   
3                    1              0               1         98.00  ...   
4                    1              0               1         83.90  ...   
...                ...            ...             ...           ...  ...   
7262                 1              1               0         55.15  ...   
7263                 1              0               1         85.10  ...   
7264                 1              0               1         50.30  ...   
7265                 1              2               0         67.85  ...   
7266                 0              2               0         59.00  ...   

      Seguranca_Online  Backup_Online  Protecao_Dispositivo  \
0                    0              1                     0   
1                    0              0                     0   
2                    0              0                     1   
3                    0              1                     1   
4                    0              0                     0   
...                ...            ...                   ...   
7262                 1              0                     0   
7263                 0              0                     0   
7264                 0              1                     0   
7265                 1              0                     1   
7266                 1              1                     1   

      Suporte_Tecnico_Internet  TV_Cabo  Streaming_Filmes  \
0                            1        1                 0   
1                            0        0                 1   
2                            0        0                 0   
3                            0        1                 1   
4                            1        1                 0   
...                        ...      ...               ...   
7262                         1        0                 0   
7263                         0        0                 1   
7264                         0        0                 0   
7265                         1        0                 1   
7266                         0        1                 1   

      Transferencia_Bancaria  Cartao_Credito  Cheque_Eletronico  Cheque  
0                          0               0                  0       1  
1                          0               0                  0       1  
2                          0               0                  1       0  
3                          0               0                  1       0  
4                          0               0                  0       1  
...                      ...             ...                ...     ...  
7262                       0               0                  0       1  
7263                       0               0              

## Salvando o arquivo
O ultimo passo é salvar o arquivo tratado no formato CSV

In [ ]:
df.to_csv(".../Dados/01-Telco-Customer-Churn-dados-limpos.csv", index = False)  